drivers: sensor: bh1750: Add driver for BH1750

heitzmann opened the pull request:
Add driver for the Rohm BH1750 16 bit light sensor.  The sensor operates
in one of 3 resolution modes, selected thru Kconfig, and the driver
allows runtime configuration of the oversampling factor.

Signed-off-by: Lucas Heitzmann Gabrielli <heitzmann@gmail.com>

heitzmann committed:
drivers: sensor: bh1750: Add driver for BH1750

Add driver for the Rohm BH1750 16 bit light sensor.  The sensor operates
in one of 3 resolution modes, selected thru Kconfig, and the driver
allows runtime configuration of the oversampling factor.

Signed-off-by: Lucas Heitzmann Gabrielli <heitzmann@gmail.com>

ycsin commented on drivers/sensor/bh1750/Kconfig:
```diff
@@ -0,0 +1,29 @@
+	  with typical sensitivities of 1 lx for normal and 4 lx for low res.
+	  In high res., the illuminance per LSB is 28.75 / oversampling-factor
+	  and the typical sensitivity is 0.5 lx.
+config BH1750_NORMAL_RES
```
one empty line above this line

ycsin commented on drivers/sensor/bh1750/Kconfig:
```diff
@@ -0,0 +1,29 @@
+	  and the typical sensitivity is 0.5 lx.
+config BH1750_NORMAL_RES
+	bool "normal resolution"
+config BH1750_HIGH_RES
```
one empty line above this line

ycsin commented on drivers/sensor/bh1750/Kconfig:
```diff
@@ -0,0 +1,29 @@
+	bool "normal resolution"
+config BH1750_HIGH_RES
+	bool "high resolution"
+config BH1750_LOW_RES
```
one empty line above this line

ycsin commented on drivers/sensor/bh1750/bh1750.c:
```diff
@@ -0,0 +1,153 @@
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_LIGHT || chan == SENSOR_CHAN_ALL);
+	__ASSERT_NO_MSG(attr == SENSOR_ATTR_OVERSAMPLING);
+
+	if (val->val1 < 31 || val->val1 > 254) {
```
use a macro instead of magic number

ycsin commented on drivers/sensor/bh1750/bh1750.c:
```diff
@@ -0,0 +1,153 @@
+	byte = BH1750_MEAS_TIME_MSB | (((uint8_t)val->val1 >> 5) & 0x07);
+	ret = bh1750_write(dev, byte);
+	if (ret < 0) {
+		LOG_ERR("Failed to configure high part of oversampling factor.");
```
```suggestion
		LOG_ERR("Failed to configure %s part of oversampling factor.", "high");
```

You can save a few bytes if you do it like so, same goes for the low part below

ycsin commented on drivers/sensor/bh1750/bh1750.c:
```diff
@@ -0,0 +1,153 @@
+		LOG_ERR("Unable to issue read command.");
+		return ret;
+	}
+	data->adc_count = ((uint16_t)buffer[0] << 8) | (uint16_t)buffer[1];
```
theres a function for this
https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/sys/byteorder.h#L377

ycsin commented on drivers/sensor/bh1750/bh1750.c:
```diff
@@ -0,0 +1,153 @@
+{
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_LIGHT);
+	struct bh1750_data *data = dev->data;
+	uint32_t tmp = (uint32_t)data->adc_count * (230 << 8) / data->oversampling_factor;
```
one empty line after var declaration

ycsin commented on drivers/sensor/bh1750/bh1750.c:
```diff
@@ -0,0 +1,153 @@
+#if defined(CONFIG_BH1750_HIGH_RES)
+	/* tmp = lux * 2^11 */
+	val->val1 = (int32_t)(tmp >> 11);
+	val->val2 = (int32_t)((tmp & 0x07FF) * 15625) >> 5;
```
use macro to better document these magic numbers?

heitzmann replied:
The numbers here come from a mix of the datasheet equations and shifts to preserve the most significant digits without increasing computational cost (or risking overflow). I couldn't think of macro names that would properly document them, so I added actual documentation to the function, if that's okay.

ycsin replied:
Is there a name for the bit/register position/mask?

heitzmann replied:
No, the datasheet doesn't describe the registers as usual. It only mentions the data register (that holds the last ADC count, which is the name I used) and the measurement time register (which I renamed to oversampling factor to keep in line with the zephyr convention). The IC is very simple; I don't believe there are any other registers.

ycsin commented on drivers/sensor/bh1750/bh1750.h:
```diff
@@ -0,0 +1,45 @@
+
+#if defined(CONFIG_BH1750_NORMAL_RES)
+#define BH1750_SINGLE_MEASUREMENT BH1750_CMD_HRES
+#define BH1750_TIME_FACTOR ((uint32_t)2610)
```
I think you only need `2610U`?

ycsin commented on dts/bindings/sensor/rohm,bh1750-i2c.yaml:
```diff
@@ -0,0 +1,21 @@
+      Acceptable values are from 31 to 254, corresponding to (0.45 and 3.68
+      times the default value, respectively).  The sampling time is this factor
+      multiplied by 2.61 ms (normal and high resolution) or 0.35 ms (low res).
+    default: 69
```
You might have to reason why 69 is the default

heitzmann replied:
I've improved the description in more accurate terms and included the reason for the default value (it is the reset default).

galak commented on drivers/sensor/bh1750/Kconfig:
```diff
@@ -0,0 +1,29 @@
+
+menuconfig BH1750
+	bool "Rohm BH1750 16 bit ambient light sensor"
+	depends on I2C
```
add:
```
default y
depends on DT_HAS_ROHM_BH1750_ENABLED
```

MaureenHelm replied:
This still needs to be done

galak commented on dts/bindings/sensor/rohm,bh1750-i2c.yaml:
```diff
@@ -0,0 +1,21 @@
+# Copyright (c) 2022 Lucas Heitzmann Gabrielli
```
rename file to just `rohm,bh1750.yaml`.  We use the `-i2c` when a device supports different buses like `i2c` and  `spi`.

galak reviewed (changes requested):
* Kconfig update
* rename binding yaml file.

heitzmann committed:
drivers: sensor: bh1759: Improve src documentation

Include review sugestions and code compliance fixes.

Signed-off-by: Lucas Heitzmann Gabrielli <heitzmann@gmail.com>

heitzmann commented:
Thanks @galak and @ycsin for the reviews and comments! I believe I've implemented all the changes.

heitzmann committed:
drivers: sensor: bh1750: Fix includes

Add missing include for sys_get_be16 and prefix others with zephyr root.

Signed-off-by: Lucas Heitzmann Gabrielli <heitzmann@gmail.com>

galak commented on drivers/sensor/bh1750/bh1750.h:
```diff
@@ -0,0 +1,48 @@
+#define BH1750_MEAS_TIME_MAX 254
+
+struct bh1750_config {
+	const struct device *bus;
```
Make this use `struct i2c_dt_spec i2c;` instead.

See `drivers/sensor/adt7420` for an example

galak commented on drivers/sensor/bh1750/bh1750.c:
```diff
@@ -0,0 +1,184 @@
+
+#define BH1750_INST(inst)                                                                          \
+	static const struct bh1750_config bh1750_config_##inst = {                                 \
+		.bus = DEVICE_DT_GET(DT_INST_BUS(inst)),                                           \
```
replace with `I2C_DT_SPEC_INST_GET()` and `struct i2c_dt_spec` usage.

galak commented on drivers/sensor/bh1750/bh1750.c:
```diff
@@ -0,0 +1,184 @@
+{
+	const struct bh1750_config *cfg = dev->config;
+
+	return i2c_read(cfg->bus, byte, count, cfg->bus_addr);
```
use i2c_dt_spec and i2c_read_dt

galak commented on drivers/sensor/bh1750/bh1750.c:
```diff
@@ -0,0 +1,184 @@
+{
+	const struct bh1750_config *cfg = dev->config;
+
+	return i2c_write(cfg->bus, &byte, 1, cfg->bus_addr);
```
use i2c_dt_spec and i2c_write_dt

galak reviewed (changes requested):
Update to use i2c_dt_spec and associated APIs

heitzmann committed:
drivers: sensor: bh1750: Use i2c_dt_spec interface

Store i2c spec in the instance config instead of bus and address
directly.  Update i2c read/write calls to use the interface.

Signed-off-by: Lucas Heitzmann Gabrielli <heitzmann@gmail.com>

heitzmann committed:
drivers: sensor: bh1750: Add missing include i2c.h

Also remove unnecessary headers and include stdint explicitly.

Signed-off-by: Lucas Heitzmann Gabrielli <heitzmann@gmail.com>

morsisko commented:
I see that we kinda did the same job twice: https://github.com/zephyrproject-rtos/zephyr/pull/43086

MaureenHelm commented on drivers/sensor/bh1750/Kconfig:
```diff
@@ -0,0 +1,37 @@
+if BH1750
+
+choice
+	prompt "BH1750 resolution mode"
```
This option should be converted to a devicetree property to allow configuring it on a per-instance basis

MaureenHelm commented on drivers/sensor/bh1750/bh1750.c:
```diff
@@ -0,0 +1,185 @@
+
+static int bh1750_sample_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_LIGHT);
```
Return an errno

MaureenHelm commented on drivers/sensor/bh1750/bh1750.c:
```diff
@@ -0,0 +1,185 @@
+static int bh1750_channel_get(const struct device *dev, enum sensor_channel chan,
+			      struct sensor_value *val)
+{
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_LIGHT);
```
Return an errno

MaureenHelm commented on dts/bindings/sensor/rohm,bh1750.yaml:
```diff
@@ -0,0 +1,24 @@
+      2.61 ms * oversampling-factor for high and normal resolution modes, and
+      as 0.35 ms * oversampling-factor for low resolution. The minimal and
+      maximal values are 31 and 254.
+    default: 69
```
Why was this default value chosen? Is it a power-on register reset value?

MaureenHelm reviewed (changes requested):
Please squash the fixup commits into the original commit
